File: Story1.cpp
/*
Author     : Avani Shrimali
Program    : Week 1 Stories – OOP & Problem Solving Tasks
Language   : C++ 
Description:
    This program implements seven independent problems.

    Problems Implemented:
      1. Efficient Traveler        – Binary feasibility check
      2. Sum That Stands Out       – Prefix/Suffix sum balance
      3. Flip the Switch           – Binary group flip counter
      4. Odd One Out               – XOR-based unique finder
      5. Smart Pair Finder         – Brute-force pair search
      6. Minimalist Painter        – Skip-one minimum total
      7. Geometry Problem          – OOP hierarchy demo (Point,
                                     Circle, Square, inheritance)

Error Handling:
    - Every numeric input validated using IsValidInteger().
    - cin.fail() checks prevent crashes on bad inputs.
    - All loops and arrays guarded by size limits.
    - Program exits gracefully without undefined behavior.

*/

#include <iostream>
#include <cmath>
using namespace std;

// Utility Class: for helper functions
class Utility {
public:
    // Validate integer string (handles negatives)
    bool IsValidInteger(const char input[]) {
        if (input[0] == '\0') return false;
        int i = (input[0] == '-') ? 1 : 0;
        for (; input[i] != '\0'; ++i)
            if (input[i] < '0' || input[i] > '9')
                return false;
        return true;
    }

    // Read integer array safely (space-separated).
    // Returns the number of integers read (0..maxSize).
    int ReadIntegerArray(int arr[], int maxSize) {
        int index = 0;
        char input[64];
        cout << "Enter numbers (space separated, stop with non-integer): ";
        while (index < maxSize) {
            if (!(cin >> input)) break;
            if (!IsValidInteger(input)) break;
            arr[index++] = atoi(input);
        }
        // clear leftover from input
        cin.clear();
        cin.ignore(10000, '\n');
        return index;
    }

    // Simple helper utilities
    long long CalculatePrefixSum(int arr[], int upToIndex) {
        long long sum = 0;
        for (int i = 0; i < upToIndex; ++i) sum += arr[i];
        return sum;
    }

    long long CalculateSuffixSum(int arr[], int fromIndex, int size) {
        long long sum = 0;
        for (int i = fromIndex + 1; i < size; ++i) sum += arr[i];
        return sum;
    }

    int FindMinimum(int a, int b) { return (a < b) ? a : b; }

    int FindXOR(int arr[], int size) {
        int result = 0;
        for (int i = 0; i < size; ++i) result ^= arr[i];
        return result;
    }
};

// Story 1
class ProblemSolver {
private:
    Utility util;
public:
    // 1. Efficient Traveler (fixed)
    // energyCosts: array of n costs, initialEnergy: starting energy
    // returns farthest city index reachable (0-based). If no cities given returns -1.
    int FindFarthestCity(int energyCosts[], int n, int initialEnergy) {
        if (n <= 0) return -1;            // no cities given
        long long currentEnergy = initialEnergy;
        for (int i = 0; i < n; ++i) {
            if (currentEnergy - energyCosts[i] < 0)
                return i;                 // can reach city i but not move to i+1
            currentEnergy -= energyCosts[i];
        }
        return n - 1;                     // reached last city
    }

    // 2. Sum That Stands Out
    // find index i where sum(arr[0..i-1]) == sum(arr[i+1..n-1])
    int FindBalancedIndex(int numbers[], int n) {
        for (int i = 0; i < n; ++i)
            if (util.CalculatePrefixSum(numbers, i) ==
                util.CalculateSuffixSum(numbers, i, n))
                return i;
        return -1;
    }

    // 3. Flip the Switch
    // Assumes binaryArray contains only 0s and 1s; if not, it treats other values as groups too.
    int FindMinimumFlips(int binaryArray[], int n) {
        if (n <= 0) return 0;
        int zeroGroups = 0, oneGroups = 0, i = 0;
        while (i < n) {
            int bitValue = binaryArray[i];
            int j = i;
            while (j < n && binaryArray[j] == bitValue) j++;
            if (bitValue == 0) zeroGroups++;
            else oneGroups++;
            i = j;
        }
        return util.FindMinimum(zeroGroups, oneGroups);
    }

    // 4. Odd One Out
    int FindUniqueElement(int numbers[], int n) {
        return util.FindXOR(numbers, n);
    }

    // 5. Smart Pair Finder
    bool IsPairWithSumExists(int numbers[], int n, int targetSum) {
        for (int i = 0; i < n; ++i)
            for (int j = i + 1; j < n; ++j)
                if (numbers[i] + numbers[j] == targetSum)
                    return true;
        return false;
    }

    // 6. Minimalist Painter
    // Skip exactly one wall to minimize total time. If n==0 -> 0, if n==1 -> 0.
    int FindMinimalPaintingTime(int times[], int n) {
        if (n <= 0) return 0;
        long long totalSum = 0;
        int minTime = times[0];
        for (int i = 0; i < n; ++i) {
            totalSum += times[i];
            if (times[i] < minTime) minTime = times[i];
        }
        return (int)(totalSum - minTime);
    }
};

// Problem 7.
class Point {
public:
    double x, y;
    Point(double xVal = 0, double yVal = 0) : x(xVal), y(yVal) {}
    double Distance(const Point &other) const {
        return sqrt((x - other.x) * (x - other.x) +
                    (y - other.y) * (y - other.y));
    }
    void Translate(double dx, double dy) { x += dx; y += dy; }
    void Print() const { cout << "(" << x << ", " << y << ")"; }
};

class Shape {
public:
    virtual double GetArea() const = 0;
    virtual ~Shape() {}
};

class Circle : public Shape {
private:
    double radius;
public:
    Circle(double r) : radius(r) {}
    double GetArea() const override { return 3.1416 * radius * radius; }
};

class Square : public Shape {
private:
    double side;
public:
    Square(double s) : side(s) {}
    double GetArea() const override { return side * side; }
};

int main() {
    ProblemSolver solver;
    Utility util;
    int choice;

    do {
        cout << "\nMenu\n";
        cout << "1) Efficient Traveler\n2) Sum That Stands Out\n3) Flip the Switch\n";
        cout << "4) Odd One Out\n5) Smart Pair Finder\n6) Minimalist Painter\n";
        cout << "7) Geometry Problem\n0) Exit\nChoice: ";

        cin >> choice;
        if (cin.fail()) {
            cin.clear();
            cin.ignore(10000, '\n');
            cout << "Invalid input. Please enter a number between 0 and 7.\n";
            continue;
        }

        switch (choice) {
            case 1: {
                // Efficient Traveler
                char energyStr[64];
                do { cout << "Enter initial energy: "; cin >> energyStr; }
                while (!util.IsValidInteger(energyStr));
                int initialEnergy = atoi(energyStr);

                int energyCosts[100];
                int n = util.ReadIntegerArray(energyCosts, 100);
                int farthest = solver.FindFarthestCity(energyCosts, n, initialEnergy);
                cout << "Farthest city index reachable: " << farthest << endl;
            } break;

            case 2: {
                // Sum That Stands Out
                int numbers[100];
                int n = util.ReadIntegerArray(numbers, 100);
                cout << "Balanced index: " << solver.FindBalancedIndex(numbers, n) << endl;
            } break;

            case 3: {
                // Flip the Switch
                int binaryArr[100];
                int n = util.ReadIntegerArray(binaryArr, 100);
                cout << "Minimum flips needed: " << solver.FindMinimumFlips(binaryArr, n) << endl;
            } break;

            case 4: {
                // Odd One Out
                int arr[100];
                int n = util.ReadIntegerArray(arr, 100);
                if (n <= 0) cout << "No input provided.\n";
                else cout << "Unique element: " << solver.FindUniqueElement(arr, n) << endl;
            } break;

            case 5: {
                // Smart Pair Finder
                int arr[100];
                int n = util.ReadIntegerArray(arr, 100);
                if (n <= 1) {
                    cout << "Need at least two numbers to find a pair.\n";
                    break;
                }
                char targetStr[64];
                do { cout << "Enter target sum: "; cin >> targetStr; }
                while (!util.IsValidInteger(targetStr));
                int target = atoi(targetStr);
                cout << "Pair exists? "
                     << (solver.IsPairWithSumExists(arr, n, target) ? "Yes" : "No") << endl;
            } break;

            case 6: {
                // Minimalist Painter
                int arr[100];
                int n = util.ReadIntegerArray(arr, 100);
                cout << "Minimal total painting time: "
                     << solver.FindMinimalPaintingTime(arr, n) << endl;
            } break;

            case 7: {
                // Geometry Problem
                double x1, y1, x2, y2;
                cout << "Enter x y of Point1: "; cin >> x1 >> y1;
                cout << "Enter x y of Point2: "; cin >> x2 >> y2;

                Point p1(x1, y1), p2(x2, y2);
                cout << "Distance: " << p1.Distance(p2) << endl;

                double radius, side;
                cout << "Enter circle radius: "; cin >> radius;
                cout << "Enter square side: "; cin >> side;

                Circle c(radius);
                Square s(side);
                cout << "Circle Area: " << c.GetArea() << endl;
                cout << "Square Area: " << s.GetArea() << endl;

                double dx, dy;
                cout << "Enter translation dx dy: "; cin >> dx >> dy;
                p1.Translate(dx, dy);
                cout << "Translated Point1: "; p1.Print(); cout << endl;
            } break;

            case 0:
                cout << "Exiting program.\n";
                break;

            default:
                cout << "Invalid choice. Please enter a number between 0 and 7.\n";
                break;
        }

    } while (choice != 0);

    cout << "Program ended.\n";
    return 0;
}

File: Story 1/story1.java
import java.util.InputMismatchException;
import java.util.Scanner;

/**
 * @file StoryProblems.java
 * @description Menu-driven program implementing Week 1 Stories:
 * 1. Efficient Traveler
 * 2. The Sum That Stands Out (Pivot Index)
 * 3. Flip The Switch
 * 4. The Odd One Out (Manual Frequency Array)
 * 5. Smart Pair Finder (Manual Frequency Array)
 * 6. The Minimalist Painter
 * @author - Badal
 */

public class story1{

    //  Efficient Traveler
    public int efficientTraveler(int[] nums, int initialEnergy) {
        int tempEnergy = initialEnergy;
        for (int i = 0; i < nums.length; i++) {
            tempEnergy -= nums[i];
            if (tempEnergy < 0)
                return i ;
        }
        return nums.length - 1;
    }

    //  The Sum That Stands Out (Pivot Index)
    public int pivotIndex(int[] nums) {
        int totalSum = 0;
        for (int i = 0; i < nums.length; i++)
            totalSum += nums[i];
        int leftSum = 0;
        for (int i = 0; i < nums.length; i++) {
            totalSum -= nums[i];
            if (leftSum == totalSum)
                return i;
            leftSum += nums[i];
        }
        return -1;
    }

    //  Flip the Switch
    public int flips(int[] nums) {
        int count0 = 0, count1 = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0)
                count0++;
            else
                count1++;
        }
        return (count0 < count1) ? count0 : count1;
    }

    //  The Odd One Out — manual frequency array
    public int oddOneOut(int[] nums) {
        int max = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] > max)
                max = nums[i];
        }

        int[] freq = new int[max + 1];
        for (int i = 0; i < nums.length; i++) {
            freq[nums[i]]++;
        }

        for (int i = 0; i <= max; i++) {
            if (freq[i] == 1)
                return i;
        }
        return -1;
    }

    //  Smart Pair Finder — manual frequency array
    public boolean smartPairFinder(int[] nums, int target) {
       for(int i=0;i<nums.length-1;i++){
           for(int j=i+1;j<nums.length;j++){
               if(nums[i]+nums[j]==target)
                   return true;
           }
       }
       return false;
    }

    // The Minimalist Painter
    public int minimalistPainter(int[] time) {
        if (time.length == 0)
            return 0;
        int total = 0;
        int max = time[0];
        for (int i = 0; i < time.length; i++) {
            total += time[i];
            if (time[i] > max)
                max = time[i];
        }
        return total - max;
    }

    // Menu-Driven Program
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        story1 obj = new story1();

        while (true) {
            try {
                System.out.println("\n--------- Week 1 Stories Menu ---------");
                System.out.println("1. Efficient Traveler");
                System.out.println("2. The Sum That Stands Out (Pivot Index)");
                System.out.println("3. Flip The Switch");
                System.out.println("4. The Odd One Out");
                System.out.println("5. Smart Pair Finder");
                System.out.println("6. The Minimalist Painter");
                System.out.println("7. Exit");
                System.out.print("Enter your choice: ");

                int ch = sc.nextInt();

                switch (ch) {
                    case 1:
                        System.out.print("Enter number of cities: ");
                        int n1 = sc.nextInt();
                        int[] nums1 = new int[n1];
                        System.out.println("Enter energy costs:");
                        for (int i = 0; i < n1; i++)
                            nums1[i] = sc.nextInt();
                        System.out.print("Enter initial energy: ");
                        int e = sc.nextInt();
                        System.out.println("Farthest city index: " + obj.efficientTraveler(nums1, e));
                        break;

                    case 2:
                        System.out.print("Enter array size: ");
                        int n2 = sc.nextInt();
                        int[] nums2 = new int[n2];
                        System.out.println("Enter elements:");
                        for (int i = 0; i < n2; i++)
                            nums2[i] = sc.nextInt();
                        System.out.println("Pivot Index: " + obj.pivotIndex(nums2));
                        break;

                    case 3:
                        System.out.print("Enter binary array size: ");
                        int n3 = sc.nextInt();
                        int[] nums3 = new int[n3];
                        System.out.println("Enter binary elements (0/1):");
                        for (int i = 0; i < n3; i++)
                            nums3[i] = sc.nextInt();
                        System.out.println("Minimum flips: " + obj.flips(nums3));
                        break;

                    case 4:
                        System.out.print("Enter array size: ");
                        int n4 = sc.nextInt();
                        int[] nums4 = new int[n4];
                        System.out.println("Enter elements (each appears twice except one):");
                        for (int i = 0; i < n4; i++)
                            nums4[i] = sc.nextInt();
                        System.out.println("Odd one out: " + obj.oddOneOut(nums4));
                        break;

                    case 5:
                        System.out.print("Enter array size: ");
                        int n5 = sc.nextInt();
                        int[] nums5 = new int[n5];
                        System.out.println("Enter elements:");
                        for (int i = 0; i < n5; i++)
                            nums5[i] = sc.nextInt();
                        System.out.print("Enter target sum: ");
                        int target = sc.nextInt();
                        System.out.println("Pair with sum " + target + " exists: " + obj.smartPairFinder(nums5, target));
                        break;

                    case 6:
                        System.out.print("Enter number of walls: ");
                        int n6 = sc.nextInt();
                        int[] time = new int[n6];
                        System.out.println("Enter painting times:");
                        for (int i = 0; i < n6; i++)
                            time[i] = sc.nextInt();
                        System.out.println("Minimum total time (after skipping one wall): " + obj.minimalistPainter(time));
                        break;

                    case 7:
                        System.out.print("Are you sure you want to exit? (y/n): ");
                        char confirm = sc.next().charAt(0);
                        if (confirm == 'y' || confirm == 'Y') {
                            System.out.println("Exiting program. Thank you!");
                            sc.close();
                            return;
                        } else {
                            System.out.println("Exit cancelled. Returning to menu...");
                        }
                        break;

                    default:
                        System.out.println("Wrong choice! Please enter a number between 1 and 7.");
                }
            } catch (InputMismatchException e) {
                System.out.println(" Invalid input! Please enter only numeric values given in Menu.");
                sc.nextLine(); // clear invalid input
            }
        }
    }
}
